ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java

```diff
@@ -261,45 +283,151 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
     return true;
   }
 
-  // Allocate unique datanode.disks for each replica in each datacenter up to replicaCountPerDatacenter (hard-code all
-  // datacenters to have same number of replicas for now).
+  /**
+   * Get a sampling of {@code numDisks} random disks from a list of {@link DataNode}s.
+   * NOTE 1: This method will change the ordering of the nodes in {@code dataNodes}
+   * NOTE 2: If {@code numDisks} valid disks could not be found, the returned list could be shorter than
+   *         {@code numDisks}
+   *
+   * @param dataNodes the list of {@link DataNode}s to sample from
+   * @param dataNodesUsed the set of {@link DataNode}s to exclude from the sample
+   * @param replicaCapacityInBytes the minimum amount of free space that a disk in the sample should have
+   * @param rackAware if {@code true}, only return disks in nodes that do not share racks with the nodes
+   *                  in {@code dataNodesUsed}
+   * @param numDisks how many disks to return in the sample
+   * @return a list of {@link Disk}s of length {@code numDisks}, or fewer if {@code numDisks} valid samples could not
+   *         be found
+   */
+  private List<Disk> getRandomDiskCandidateSample(List<DataNode> dataNodes, Set<DataNode> dataNodesUsed,
+      long replicaCapacityInBytes, boolean rackAware, int numDisks) {
+    Set<Long> rackIdsUsed = new HashSet<>();
+    if (rackAware) {
+      for (DataNode dataNode : dataNodesUsed) {
+        rackIdsUsed.add(dataNode.getRackId());
+      }
+    }
+    List<Disk> diskCandidates = new ArrayList<>();
+    int selectionBound = dataNodes.size();
+    Random randomGen = new Random();
+    while ((selectionBound > 0) && (diskCandidates.size() < numDisks)) {
+      int selectionIndex = randomGen.nextInt(selectionBound);
+      DataNode candidate = dataNodes.get(selectionIndex);
+      if (!dataNodesUsed.contains(candidate) && !rackIdsUsed.contains(candidate.getRackId())) {
+        Disk diskWithMostCapacity = getDiskWithMostUnallocatedRawCapacity(candidate, replicaCapacityInBytes);
```

Comment:
as a general question - what is the rationale in choosing the disk with the most free space really? Basically choosing worst fit allocation over best fit allocation? I understand that this way one might allocate the least loaded disk (assuming all disks have even capacities to begin with). Setting that aside, is worst fit a better allocation strategy?